File: frontend/src/services/taskService.ts
import { apiClient } from '../api/client';

export type TaskCategoryId =
  | 'grocery'
  | 'laundry'
  | 'documents'
  | 'queue'
  | 'delivery'
  | 'tutoring'
  | 'cleaning'
  | 'tech'
  | 'moving'
  | 'other';

export type TaskStatus = 'open' | 'bidding' | 'assigned' | 'in_progress' | 'completed' | 'cancelled' | 'disputed';

export type BidStatus = 'pending' | 'accepted' | 'rejected' | 'withdrawn';

export interface UserSummary {
  id: string;
  fullName: string;
  avatarUrl: string | null;
  verified: boolean;
}

export interface TaskItem {
  id: string;
  title: string;
  description: string;
  category: TaskCategoryId;
  budget: number;
  location: string;
  locationLat: number | null;
  locationLng: number | null;
  isDelivery: boolean;
  pickupLocation: string | null;
  pickupLat: number | null;
  pickupLng: number | null;
  dropoffLocation: string | null;
  dropoffLat: number | null;
  dropoffLng: number | null;
  deadline: string;
  isUrgent: boolean;
  status: TaskStatus;
  poster: UserSummary;
  assignedTasker: UserSummary | null;
  finalPrice: number | null;
  bidCount: number;
  myBidStatus: BidStatus | null;
  createdAt: string;
  completedAt: string | null;
}

export interface BidItem {
  id: string;
  taskId: string;
  tasker: UserSummary;
  amount: number;
  message: string | null;
  deliveryTime: string | null;
  status: BidStatus;
  createdAt: string;
}

export interface CreateTaskPayload {
  title: string;
  description: string;
  category: TaskCategoryId;
  budget: number;
  location: string;
  locationLat?: number | null;
  locationLng?: number | null;
  isDelivery: boolean;
  pickupLocation?: string | null;
  pickupLat?: number | null;
  pickupLng?: number | null;
  dropoffLocation?: string | null;
  dropoffLat?: number | null;
  dropoffLng?: number | null;
  deadline: string;
  isUrgent: boolean;
}

export interface PlaceBidPayload {
  amount: number;
  message?: string;
  deliveryTime?: string;
}

export async function createTask(payload: CreateTaskPayload): Promise<TaskItem> {
  const res = await apiClient.post<TaskItem>('/api/tasks', payload);
  return res.data;
}

export async function getOpenTasks(params?: { category?: string; search?: string }): Promise<TaskItem[]> {
  const res = await apiClient.get<TaskItem[]>('/api/tasks', { params });
  return res.data;
}

export async function getMyTasks(role: 'posted' | 'assigned'): Promise<TaskItem[]> {
  const res = await apiClient.get<TaskItem[]>('/api/tasks/mine', { params: { role } });
  return res.data;
}

export async function getTask(id: string): Promise<TaskItem> {
  const res = await apiClient.get<TaskItem>(`/api/tasks/${id}`);
  return res.data;
}

export async function acceptTask(id: string): Promise<TaskItem> {
  const res = await apiClient.post<TaskItem>(`/api/tasks/${id}/accept`);
  return res.data;
}

export async function cancelTask(id: string): Promise<TaskItem> {
  const res = await apiClient.post<TaskItem>(`/api/tasks/${id}/cancel`);
  return res.data;
}

export async function completeTask(id: string): Promise<TaskItem> {
  const res = await apiClient.post<TaskItem>(`/api/tasks/${id}/complete`);
  return res.data;
}

export async function placeBid(taskId: string, payload: PlaceBidPayload): Promise<BidItem> {
  const res = await apiClient.post<BidItem>(`/api/tasks/${taskId}/bids`, payload);
  return res.data;
}

export async function getBidsForTask(taskId: string): Promise<BidItem[]> {
  const res = await apiClient.get<BidItem[]>(`/api/tasks/${taskId}/bids`);
  return res.data;
}

export async function acceptBid(bidId: string): Promise<TaskItem> {
  const res = await apiClient.post<TaskItem>(`/api/bids/${bidId}/accept`);
  return res.data;
}

export async function withdrawBid(bidId: string): Promise<BidItem> {
  const res = await apiClient.post<BidItem>(`/api/bids/${bidId}/withdraw`);
  return res.data;
}

export interface ConversationSummary {
  id: string;
  otherUser: UserSummary;
  relatedTaskId: string | null;
  lastMessage: string | null;
  lastMessageAt: string | null;
  unreadCount: number;
  createdAt: string;
}

export async function startConversationWith(userId: string, taskId?: string): Promise<ConversationSummary> {
  const res = await apiClient.post<ConversationSummary>(`/api/conversations/with/${userId}`, taskId ? { taskId } : {});
  return res.data;
}

export async function bookmarkTask(taskId: string): Promise<void> {
  await apiClient.post(`/api/tasks/${taskId}/bookmark`);
}

export async function unbookmarkTask(taskId: string): Promise<void> {
  await apiClient.delete(`/api/tasks/${taskId}/bookmark`);
}

export async function getBookmarkStatus(taskId: string): Promise<boolean> {
  const res = await apiClient.get<{ bookmarked: boolean }>(`/api/tasks/${taskId}/bookmark`);
  return res.data.bookmarked;
}

export async function getBookmarkedTasks(): Promise<TaskItem[]> {
  const res = await apiClient.get<TaskItem[]>('/api/tasks/bookmarks');
  return res.data;
}

export interface TaskStatusUpdateItem {
  id: string;
  note: string;
  author: UserSummary;
  createdAt: string;
}

export async function getTaskStatusUpdates(taskId: string): Promise<TaskStatusUpdateItem[]> {
  const res = await apiClient.get<TaskStatusUpdateItem[]>(`/api/tasks/${taskId}/status-updates`);
  return res.data;
}

export async function postTaskStatusUpdate(taskId: string, note: string): Promise<TaskStatusUpdateItem> {
  const res = await apiClient.post<TaskStatusUpdateItem>(`/api/tasks/${taskId}/status-updates`, { note });
  return res.data;
}

File: services/tasks-service/src/main/java/com/hustlehub/tasks/service/TaskService.java
package com.hustlehub.tasks.service;

import com.hustlehub.common.client.NotificationsServiceClient;
import com.hustlehub.common.client.PaymentsServiceClient;
import com.hustlehub.common.client.UserServiceClient;
import com.hustlehub.common.dto.CompletedEngagementResponse;
import com.hustlehub.common.dto.EngagementParticipantsResponse;
import com.hustlehub.common.dto.NotificationType;
import com.hustlehub.common.dto.TransferResult;
import com.hustlehub.common.dto.UserStatsResponse;
import com.hustlehub.common.dto.UserSummaryResponse;
import com.hustlehub.common.exception.ForbiddenActionException;
import com.hustlehub.common.exception.InvalidRequestException;
import com.hustlehub.common.exception.ResourceNotFoundException;
import com.hustlehub.tasks.dto.request.CreateTaskRequest;
import com.hustlehub.tasks.dto.response.TaskResponse;
import com.hustlehub.tasks.entity.Bid;
import com.hustlehub.tasks.entity.BidStatus;
import com.hustlehub.tasks.entity.Task;
import com.hustlehub.tasks.entity.TaskCategory;
import com.hustlehub.tasks.entity.TaskStatus;
import com.hustlehub.tasks.exception.InvalidTaskStateException;
import com.hustlehub.tasks.exception.TaskAlreadyAssignedException;
import com.hustlehub.tasks.repository.BidRepository;
import com.hustlehub.tasks.repository.TaskRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class TaskService {

    private final TaskRepository taskRepository;
    private final BidRepository bidRepository;
    private final UserServiceClient userServiceClient;
    private final PaymentsServiceClient paymentsServiceClient;
    private final NotificationsServiceClient notificationsServiceClient;

    @Transactional
    public TaskResponse createTask(UUID posterId, CreateTaskRequest request) {
        if (request.deadline().isBefore(Instant.now())) {
            throw new InvalidTaskStateException("Deadline must be in the future");
        }
        if (request.isDelivery()) {
            boolean missingPickup = request.pickupLocation() == null || request.pickupLocation().isBlank();
            boolean missingDropoff = request.dropoffLocation() == null || request.dropoffLocation().isBlank();
            if (missingPickup || missingDropoff) {
                throw new InvalidTaskStateException("Pickup and dropoff locations are required for delivery tasks");
            }
        }

        UUID taskId = UUID.randomUUID();
        TransferResult hold = paymentsServiceClient.hold(posterId, request.budget(), "TASK", taskId,
                "Task escrow: " + request.title().trim());
        if (!hold.success()) {
            throw new InvalidRequestException("Insufficient wallet balance to post this task. Top up your wallet and try again.");
        }

        Task task = Task.builder()
                .id(taskId)
                .posterId(posterId)
                .title(request.title().trim())
                .description(request.description().trim())
                .category(TaskCategory.fromJson(request.category()))
                .budget(request.budget())
                .location(request.location().trim())
                .locationLat(request.locationLat())
                .locationLng(request.locationLng())
                .delivery(request.isDelivery())
                .pickupLocation(request.pickupLocation())
                .pickupLat(request.pickupLat())
                .pickupLng(request.pickupLng())
                .dropoffLocation(request.dropoffLocation())
                .dropoffLat(request.dropoffLat())
                .dropoffLng(request.dropoffLng())
                .deadline(request.deadline())
                .urgent(request.isUrgent())
                .build();
        task = taskRepository.save(task);
        UserSummaryResponse poster = resolveRequiredSummary(posterId);
        return TaskResponse.from(task, 0, null, poster, null);
    }

    public List<TaskResponse> getOpenTasks(UUID currentUserId, String category, String search) {
        TaskCategory categoryFilter = (category == null || category.isBlank()) ? null : TaskCategory.fromJson(category);
        String searchFilter = (search == null || search.isBlank()) ? null : search.trim().toLowerCase();

        List<Task> tasks = taskRepository.findByStatusAndPosterIdNotOrderByCreatedAtDesc(TaskStatus.OPEN, currentUserId).stream()
                .filter(task -> categoryFilter == null || task.getCategory() == categoryFilter)
                .filter(task -> searchFilter == null || task.getTitle().toLowerCase().contains(searchFilter))
                .toList();

        Map<UUID, UserSummaryResponse> summaries = resolveSummaries(tasks);
        return tasks.stream()
                .map(task -> TaskResponse.from(task, bidRepository.countByTask(task), myBidStatus(task, currentUserId),
                        summaries.get(task.getPosterId()), resolveAssignedTasker(task, summaries)))
                .toList();
    }

    public List<TaskResponse> getMyTasks(UUID currentUserId, String role) {
        List<Task> tasks = "assigned".equalsIgnoreCase(role)
                ? taskRepository.findByAssignedTaskerIdOrderByCreatedAtDesc(currentUserId)
                : taskRepository.findByPosterIdOrderByCreatedAtDesc(currentUserId);
        Map<UUID, UserSummaryResponse> summaries = resolveSummaries(tasks);
        return tasks.stream()
                .map(task -> TaskResponse.from(task, bidRepository.countByTask(task), myBidStatus(task, currentUserId),
                        summaries.get(task.getPosterId()), resolveAssignedTasker(task, summaries)))
                .toList();
    }

    public TaskResponse getTask(UUID id, UUID currentUserId) {
        Task task = findTaskOrThrow(id);
        UserSummaryResponse poster = resolveRequiredSummary(task.getPosterId());
        UserSummaryResponse assignedTasker = task.getAssignedTaskerId() != null
                ? resolveRequiredSummary(task.getAssignedTaskerId()) : null;
        return TaskResponse.from(task, bidRepository.countByTask(task), myBidStatus(task, currentUserId), poster, assignedTasker);
    }

    @Transactional
    public TaskResponse acceptTask(UUID id, UUID currentUserId) {
        Task task = findTaskOrThrow(id);
        if (task.getPosterId().equals(currentUserId)) {
            throw new ForbiddenActionException("You cannot accept your own task");
        }
        int updated = taskRepository.claimTaskAtListedPrice(id, currentUserId);
        if (updated == 0) {
            throw new TaskAlreadyAssignedException();
        }
        Task refreshed = findTaskOrThrow(id);
        UserSummaryResponse poster = resolveRequiredSummary(refreshed.getPosterId());
        UserSummaryResponse assignedTasker = resolveRequiredSummary(refreshed.getAssignedTaskerId());
        notificationsServiceClient.notify(refreshed.getPosterId(), NotificationType.TASK_ACCEPTED,
                "Your task was accepted",
                assignedTasker.fullName() + " accepted \"" + refreshed.getTitle() + "\"",
                refreshed.getId());
        return TaskResponse.from(refreshed, bidRepository.countByTask(refreshed), myBidStatus(refreshed, currentUserId), poster, assignedTasker);
    }

    @Transactional
    public TaskResponse cancelTask(UUID id, UUID currentUserId) {
        Task task = findTaskOrThrow(id);
        requirePoster(task, currentUserId);
        if (task.getStatus() != TaskStatus.OPEN) {
            throw new InvalidTaskStateException("Only open tasks can be cancelled");
        }
        task.setStatus(TaskStatus.CANCELLED);
        task = taskRepository.save(task);
        paymentsServiceClient.refundHold(task.getId());
        UserSummaryResponse poster = resolveRequiredSummary(task.getPosterId());
        return TaskResponse.from(task, bidRepository.countByTask(task), null, poster, null);
    }

    @Transactional
    public TaskResponse completeTask(UUID id, UUID currentUserId) {
        Task task = findTaskOrThrow(id);
        boolean isAssignedTasker = task.getAssignedTaskerId() != null && task.getAssignedTaskerId().equals(currentUserId);
        if (!isAssignedTasker) {
            throw new ForbiddenActionException("Only the assigned tasker can mark this task complete");
        }
        if (task.getStatus() != TaskStatus.IN_PROGRESS) {
            throw new InvalidTaskStateException("Only in-progress tasks can be marked complete");
        }
        task.setStatus(TaskStatus.COMPLETED);
        task.setCompletedAt(Instant.now());
        task = taskRepository.save(task);
        paymentsServiceClient.releaseHold(task.getId(), task.getAssignedTaskerId());

        notificationsServiceClient.notify(task.getPosterId(), NotificationType.TASK_COMPLETED,
                "Task marked complete",
                "\"" + task.getTitle() + "\" was marked complete.",
                task.getId());
        BigDecimal amountPaid = task.getFinalPrice() != null ? task.getFinalPrice() : task.getBudget();
        notificationsServiceClient.notify(task.getAssignedTaskerId(), NotificationType.TASK_PAYMENT_RECEIVED,
                "You got paid!",
                "You received GH₵" + amountPaid + " for \"" + task.getTitle() + "\"",
                task.getId());

        UserSummaryResponse poster = resolveRequiredSummary(task.getPosterId());
        UserSummaryResponse assignedTasker = task.getAssignedTaskerId() != null
                ? resolveRequiredSummary(task.getAssignedTaskerId()) : null;
        return TaskResponse.from(task, bidRepository.countByTask(task), myBidStatus(task, currentUserId), poster, assignedTasker);
    }

    Task findTaskOrThrow(UUID id) {
        return taskRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Task not found"));
    }

    void requirePoster(Task task, UUID userId) {
        if (!task.getPosterId().equals(userId)) {
            throw new ForbiddenActionException("Only the task's poster can do this");
        }
    }

    private String myBidStatus(Task task, UUID currentUserId) {
        Optional<Bid> myLatestBid = bidRepository.findByTaskAndTaskerId(task, currentUserId).stream()
                .filter(b -> b.getStatus() != BidStatus.WITHDRAWN)
                .findFirst();
        return myLatestBid.map(b -> b.getStatus().toJson()).orElse(null);
    }

    private UserSummaryResponse resolveRequiredSummary(UUID userId) {
        return userServiceClient.getSummary(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + userId));
    }

    /** Batches every poster/assignedTasker id across the whole list into a single identity-service call. */
    private Map<UUID, UserSummaryResponse> resolveSummaries(List<Task> tasks) {
        Set<UUID> ids = new HashSet<>();
        for (Task task : tasks) {
            ids.add(task.getPosterId());
            if (task.getAssignedTaskerId() != null) {
                ids.add(task.getAssignedTaskerId());
            }
        }
        return userServiceClient.getSummaries(ids).stream()
                .collect(Collectors.toMap(UserSummaryResponse::id, s -> s));
    }

    private UserSummaryResponse resolveAssignedTasker(Task task, Map<UUID, UserSummaryResponse> summaries) {
        return task.getAssignedTaskerId() == null ? null : summaries.get(task.getAssignedTaskerId());
    }

    /** For identity-service's public/own profile response — completed jobs as tasker + what they earned. */
    public UserStatsResponse getStats(UUID taskerId) {
        long completedCount = taskRepository.countByAssignedTaskerIdAndStatus(taskerId, TaskStatus.COMPLETED);
        BigDecimal totalEarnings = taskRepository.sumFinalPriceByAssignedTaskerAndStatus(taskerId, TaskStatus.COMPLETED);
        return new UserStatsResponse(completedCount, totalEarnings);
    }

    /** For reviews-service's "eligible to review" list — every completed task this user was either side of. */
    public List<CompletedEngagementResponse> getCompletedEngagements(UUID userId) {
        List<CompletedEngagementResponse> engagements = new ArrayList<>();
        for (Task task : taskRepository.findByPosterIdAndStatus(userId, TaskStatus.COMPLETED)) {
            engagements.add(new CompletedEngagementResponse("TASK", task.getId(), task.getAssignedTaskerId(), task.getCompletedAt()));
        }
        for (Task task : taskRepository.findByAssignedTaskerIdAndStatus(userId, TaskStatus.COMPLETED)) {
            engagements.add(new CompletedEngagementResponse("TASK", task.getId(), task.getPosterId(), task.getCompletedAt()));
        }
        return engagements;
    }

    /**
     * Admin-suspend cleanup: cancels (with escrow refund) every OPEN task this user posted, same
     * lifecycle transition as a self-service cancelTask, just triggered by an admin action instead
     * of the poster. Tasks already IN_PROGRESS/COMPLETED are untouched — those involve another
     * user who shouldn't be penalized by this poster's suspension.
     */
    @Transactional
    public void suspendCleanup(UUID posterId) {
        for (Task task : taskRepository.findByPosterIdAndStatus(posterId, TaskStatus.OPEN)) {
            task.setStatus(TaskStatus.CANCELLED);
            taskRepository.save(task);
            paymentsServiceClient.refundHold(task.getId());
        }
    }

    /** For reviews-service's server-side review-eligibility re-validation. Empty if the task was never assigned (no real engagement to review). */
    public Optional<EngagementParticipantsResponse> getParticipants(UUID taskId) {
        Task task = taskRepository.findById(taskId).orElse(null);
        if (task == null || task.getAssignedTaskerId() == null) {
            return Optional.empty();
        }
        return Optional.of(new EngagementParticipantsResponse(task.getPosterId(), task.getAssignedTaskerId(), task.getStatus() == TaskStatus.COMPLETED));
    }
}
